### ISSUE TITLE
MAILMAN_HOST_IP cannot be set to hostname (breaks hyperkitty logging), but that is required for more flexible dynamic container IP setups

### ISSUE THREAD

**Issue by JonasT** [maxking/docker-mailman#133]
We are currently seeing this issue with the docker containers: Mailing List archives get created in hyperkitty, but they are always empty and no mails seem to ever get added.

We did set `HYPERKITTY_API_KEY`, we did not set `HYPERKITTY_URL` since this isn't explained in the documentation for `mailman-core` at all - there is a warning emitted about this which says it's `http://mailman-web:8000/hyperkitty` right now, which is the correct internal hostname for our docker-mailman/web container, but is the port correct? I have no idea since I don't know what it is supposed to be..

So in conclusion, I'm not sure if it's a configuration issue or some other sort of bug. If it's a config issue related to `HYPERKITTY_URL`, it'd be nice if that option was properly explained in the core container readme.

Edit: I forgot to mention: no warnings or errors in the core or web container logs, apart from the "HYPERKITTY_URL not set, using the default value of http://mailman-web:8000/hyperkitty" notice

**Comment by JonasT** [comment]
To clarify, with always empty I mean e-mails we sent successfully through the list didn't appear yet after multiple hours. If that is for some reason intended, then maybe this is not a bug after all. I just feels like the archive is broken, but I am using mailman 3 for the first time so I'm not sure what the correct behavior would be (e.g. is a long delay for archiving normal?)

**Comment by maxking** [comment]
Containers don't log everything to stdout, this is probably something that should be changed though. For Django(mailman-web), I *think* it should not be hard, but `mailman-core` cannot log to stdout because of the way logging works in it.

For your issue, can you please check `/opt/mailman/core/var/logs/mailman.log` for logs of `mailman-core` and `/opt/mailman/web/logs/mailmanweb.log` & `/opt/mailman/web/logs/uwsgi.log`.

Port 8000 is correct, however, if you modified `ALLOWED_HOSTS`, it is possible you missed `mailman-web` there which would mean that it doesn't accept requests on the default `HYPERKITYY_URL`

**Comment by JonasT** [comment]
We did modify `ALLOWED_HOSTS` to include `"*"` so django access shouldn't be an issue (hopefully).

Indeed I did find something suspicious in `/opt/mailman/var/logs/mailman.log` in the `mailman-core` container:
```
Sep 29 15:27:05 2017 (28) Connection to HyperKitty failed: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
Sep 29 15:27:05 2017 (28) Exception in the HyperKitty archiver: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
Sep 29 15:27:05 2017 (28) Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/urllib3/response.py", line 302, in _error_catcher
    yield
  File "/usr/local/lib/python3.6/site-packages/urllib3/response.py", line 384, in read
    data = self._fp.read(amt)
  File "/usr/local/lib/python3.6/http/client.py", line 449, in read
    n = self.readinto(b)
  File "/usr/local/lib/python3.6/http/client.py", line 493, in readinto
    n = self.fp.readinto(b)
  File "/usr/local/lib/python3.6/socket.py", line 586, in readinto
    return self._sock.recv_into(b)
ConnectionResetError: [Errno 104] Connection reset by peer

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/requests/models.py", line 745, in generate
    for chunk in self.raw.stream(chunk_size, decode_content=True):
  File "/usr/local/lib/python3.6/site-packages/urllib3/response.py", line 436, in stream
    data = self.read(amt=amt, decode_content=decode_content)
  File "/usr/local/lib/python3.6/site-packages/urllib3/response.py", line 401, in read
    raise IncompleteRead(self._fp_bytes_read, self.length_remaining)
  File "/usr/local/lib/python3.6/contextlib.py", line 99, in __exit__
    self.gen.throw(type, value, traceback)
  File "/usr/local/lib/python3.6/site-packages/urllib3/response.py", line 320, in _error_catcher
    raise ProtocolError('Connection broken: %r' % e, e)
urllib3.exceptions.ProtocolError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/mailman_hyperkitty/__init__.py", line 154, in _archive_message
    url = self._send_message(mlist, msg)
  File "/usr/local/lib/python3.6/site-packages/mailman_hyperkitty/__init__.py", line 201, in _send_message
    files={"message": ("message.txt", message_text)})
  File "/usr/local/lib/python3.6/site-packages/requests/api.py", line 112, in post
    return request('post', url, data=data, json=json, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/api.py", line 58, in request
    return session.request(method=method, url=url, **kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 508, in request
    resp = self.send(prep, **send_kwargs)
  File "/usr/local/lib/python3.6/site-packages/requests/sessions.py", line 658, in send
    r.content
  File "/usr/local/lib/python3.6/site-packages/requests/models.py", line 823, in content
    self._content = bytes().join(self.iter_content(CONTENT_CHUNK_SIZE)) or bytes()
  File "/usr/local/lib/python3.6/site-packages/requests/models.py", line 748, in generate
    raise ChunkedEncodingError(e)
requests.exceptions.ChunkedEncodingError: ("Connection broken: ConnectionResetError(104, 'Connection reset by peer')", ConnectionResetError(104, 'Connection reset by peer'))
Sep 29 15:27:06 2017 (28) HyperKitty failure on http://mailman-web:8000/hyperkitty/api/mailman/archive: <html><title>Forbidden</title><body>
                <h1>Access is forbidden</h1></body></html> (403)
Sep 29 15:27:06 2017 (28) Exception in the HyperKitty archiver: <html><title>Forbidden</title><body>
                <h1>Access is forbidden</h1></body></html>
Sep 29 15:27:06 2017 (28) Traceback (most recent call last):
  File "/usr/local/lib/python3.6/site-packages/mailman_hyperkitty/__init__.py", line 154, in _archive_message
    url = self._send_message(mlist, msg)
  File "/usr/local/lib/python3.6/site-packages/mailman_hyperkitty/__init__.py", line 210, in _send_message
    raise ValueError(result.text)
ValueError: <html><title>Forbidden</title><body>
                <h1>Access is forbidden</h1></body></html>
```
However, `HYPERKITTY_API_KEY` is definitely set both in `-web` and `-core`, and it has the same value in both.

**Comment by JonasT** [comment]
Here is `/opt/mailman-web-data/logs`:

```
ERROR 2017-09-29 15:27:05,521 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
ERROR 2017-09-29 15:27:05,609 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
ERROR 2017-09-29 15:27:05,696 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
ERROR 2017-09-29 15:27:05,780 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
ERROR 2017-09-29 15:27:05,865 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
ERROR 2017-09-29 15:27:05,938 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
ERROR 2017-09-29 15:27:06,008 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
INFO 2017-09-29 16:00:01,211 1121 hyperkitty.lib.mailman Imported the new list <<REDACTED>> from Mailman
```
As you can see, for some reason list creation seems to work but other API access doesn't..

**Comment by JonasT** [comment]
Alright, it seems we forgot to set `MAILMAN_ARCHIVER_FROM`.

Is there any way this can be fixed so that those errors appear in the mailman-web log? That would really make a huge difference for making those errors easier to find an fix.

**Comment by JonasT** [comment]
Ok for some reason it's still broken:

```
bash-4.3# cat /opt/mailman-web-data/logs/mailmanweb.log
[...]
ERROR 2017-09-30 00:04:42,398 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
ERROR 2017-09-30 00:04:42,463 38 hyperkitty.views.mailman Access to the archiving API endpoint was forbidden from IP 172.18.0.4, your MAILMAN_ARCHIVER_FROM setting may be misconfigured
WARNING 2017-09-30 00:04:45,753 38 django.request Not Found: /favicon.ico
WARNING 2017-09-30 00:04:45,753 38 django.request Not Found: /favicon.ico
WARNING 2017-09-30 00:04:48,203 38 django.request Not Found: /favicon.ico
WARNING 2017-09-30 00:04:48,203 38 django.request Not Found: /favicon.ico
bash-4.3# echo $MAILMAN_ARCHIVER_FROM
mailman-core
bash-4.3# nslookup $MAILMAN_ARCHIVER_FROM
nslookup: can't resolve '(null)': Name does not resolve

Name:      mailman-core
Address 1: 172.18.0.4 mailman-core.network_mail_net
bash-4.3#
```

**Comment by maxking** [comment]
https://github.com/maxking/docker-mailman/blob/master/web/mailman-web/settings.py#L343 This is the configuration for logging in Django. If you need to log everything to `stdout/stderr`, you change the handler to `["console"]` in each of them, or, a combination of file & console depending on your use case. [Django's Docs for logging](http://docs.djangoproject.com/en/dev/topics/logging).

You'd have to copy the entire LOGGING dictionary to your `settings_local.py` and then modify the bits you want. 

The way I'd do it is use a environment variable to configure redirection of all logs to `console` handler.

**Comment by JonasT** [comment]
Well I don't *need* to especially since now you pointed me to this log, but it'd be nice as a general improvement. (should probably make a separate ticket for that)

Anyway, as you can see above the archiving is still broken somehow :cry:

**Comment by maxking** [comment]
Can you add an environment variable for `mailman-web` container, `MAILMAN_HOST_IP=172.18.0.4`?

And then check if things work?

**Comment by JonasT** [comment]
@maxking yup, that fixes it. Assigning static IPs is somewhat unfavorable with our current setup though (we use a larger docker network with many containers that don't have preassigned IPs and containers aren't started in a fixed order in general, so any assigned IP could already be taken), so it would be really helpful to us if this could be fixed so that hostnames work as well.

**Comment by JonasT** [comment]
I tried to give the issue a more descriptive title now that we know what is happening

**Comment by maxking** [comment]
Thank you! I am trying to debug the reason hostnames don't work. It wasn't intentional to not allow hostnames, it *could* be something to do with Hyperkitty, but maybe not.

**Comment by maxking** [comment]
Ok, so sadly Hyperkitty only checks for `request.META["REMOTE_ADDR"]` when validating clients, which means it might be impossible to make it work with hostname for now, it should be an easy fix there though. I have opened an issue there https://gitlab.com/mailman/hyperkitty/issues/157

For now, the easiest thing to do would be to get the IP Address from `/etc/hosts`(I believe if you are using Docker, it would be populated with the hostnames and IPs of containers) or use `nslookup` or something else in Python to lookup IP from Hostname when the container starts up.

**Comment by JonasT** [comment]
Ok so this little hack would make it work at least most of the time:

```python3
MAILMAN_ARCHIVER_FROM = ('mailman-core', os.environ.get('MAILMAN_HOST_IP', '172.19.199.2'))
import re                                                  
if re.search("[a-zA-Z]", MAILMAN_ARCHIVER_FROM[1]) != None:
    import socket                                                         
    MAILMAN_ARCHIVER_FROM = socket.gethostbyname(MAILMAN_ARCHIVER_FROM[1])
```

I'm not sure how to best add this though. We tried changing multiple lines in `settings.py` before, and it seems the docker startup script will then fail to insert some value at the correct place (it seems to rely on a certain value being on a certain exact line) which broke everything.

**Comment by maxking** [comment]
You can change  `settings_local.py` to add something like this:

```python3
import socket
MAILMAN_ARCHIVER_FROM = (socket.gethostbyname('mailman-core'),)
```

Yes, mailman-web is still based on Python2.7.

**Comment by JonasT** [comment]
Ok, I managed to make a work-around now. I'll document it for anyone else searching for how to temporarily get around this issue:

```Dockerfile


FROM maxking/mailman-web:latest


# Hack to allow any sort of IP to connect to Django: ( https://github.com/maxking/docker-mailman/issues/129 )

RUN python -c "import re;contents=open(\"/opt/mailman-web/settings.py\",\"rb\").read().decode(\"utf-8\",\"replace\");contents=re.sub(r\"172\\.19\\.199\\.3\",\"*\",contents);open(\"/opt/mailman-web/settings.py\",\"w\").write(contents)"


# Hack to allow MAILMAN_HOST_IP to accept hostnames: ( https://github.com/maxking/docker-mailman/issues/133 )

RUN echo -e '\n\
import os\n\
MAILMAN_ARCHIVER_FROM = ("mailman-core", os.environ.get("MAILMAN_HOST_IP", "172.19.199.2"))\n\
import re\n\
if re.search("[a-zA-Z]", MAILMAN_ARCHIVER_FROM[1]) != None and MAILMAN_ARCHIVER_FROM[1].find(":") < 0:\n\
    import socket\n\
    MAILMAN_ARCHIVER_FROM = (MAILMAN_ARCHIVER_FROM[0], socket.gethostbyname(MAILMAN_ARCHIVER_FROM[1]))' > /opt/mailman-web/settings_local.py

```

Needless to say, both those workarounds (includes the #129 workaround) are super ugly so I'd be nice if those limitations would be addressed some time in the future so we can remove those hacks. Also, **it only works if settings_local.py isn't already mounted for additional settings** - just as a note if anyone is actually planning to use this messy hack.

**Comment by maxking** [comment]
Do you need these changes because you are not using `settings_local.py`?

**Comment by JonasT** [comment]
We don't need the `settings_local.py` for anything else, and as you can see above hacking this in as a dynamic resolve seems to be a bit tedious (at least for the solution I could come up with). So yes, that is the main reason

**Comment by maxking** [comment]
That makes sense :) I will soon try to fix this in the image.
